### PR TITLE
In-memory engine: correct under-estimate memory size usage of delete entries.

### LINKED ISSUE #17106: In-memory Engine: memory usage goes to negative number
## Bug Report

image

### What version of TiKV are you using?

built on [c3b25fe](https://github.com/tikv/tikv/pull/16868/commits/c3b25fe042e83832a07c2c6c7689e1e4624faaa5)

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #17106

What's Changed:

```commit-message
fix under-estimate memory size usage of delete entry
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/region_cache_memory_engine/src/write_batch.rs [hunk 1]
```diff
@@ -396,11 +396,17 @@ impl RangeCacheWriteBatchEntry {
...
 
     pub fn calc_put_entry_size(key: &[u8], value: &[u8]) -> usize {
-        key.len() + value.len() + ENC_KEY_SEQ_LENGTH + 2 * MEM_CONTROLLER_OVERHEAD /* one for key and one for value */
+        RangeCacheWriteBatchEntry::memory_size_required_for_key_value(key, value)
     }
 
     pub fn cal_delete_entry_size(key: &[u8]) -> usize {
-        key.len() + ENC_KEY_SEQ_LENGTH
+        // delete also has value which is an empty bytes
+        RangeCacheWriteBatchEntry::memory_size_required_for_key_value(key, b"")
```
Comment (LykxSassinator) [inline components/region_cache_memory_engine/src/write_batch.rs:404]: Recommend introducing a const for the value of delete entries.

File: components/region_cache_memory_engine/src/write_batch.rs [hunk 2]
```diff
@@ -396,11 +396,17 @@ impl RangeCacheWriteBatchEntry {
     }
 
     pub fn calc_put_entry_size(key: &[u8], value: &[u8]) -> usize {
-        key.len() + value.len() + ENC_KEY_SEQ_LENGTH + 2 * MEM_CONTROLLER_OVERHEAD /* one for key and one for value */
+        RangeCacheWriteBatchEntry::memory_size_required_for_key_value(key, value)
     }
 
     pub fn cal_delete_entry_size(key: &[u8]) -> usize {
```
Comment (LykxSassinator) [inline components/region_cache_memory_engine/src/write_batch.rs:402]:
```suggestion
    pub fn calc_delete_entry_size(key: &[u8]) -> usize {
```

File: components/region_cache_memory_engine/src/keys.rs [hunk 3]
```diff
@@ -77,6 +77,11 @@ impl InternalBytes {
         } = decode_key(other.as_slice());
         user_key == other_user_key
     }
+
+    #[inline]
+    pub fn memory_size_required(bytes_size: usize) -> usize {
```
Comment (overvenus) [inline components/region_cache_memory_engine/src/keys.rs:82]: Please add some comments about how the number adds up.

**Review by zhangjinpeng87 — APPROVED** [review]
rest LGTM

File: components/region_cache_memory_engine/src/write_batch.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/write_batch.rs]: Done

File: components/region_cache_memory_engine/src/keys.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/keys.rs:85]: Done

**Review by overvenus — APPROVED** [review]

**Pull request merged**